### ISSUE TITLE
Polarization angle of 'xcal' calibrator 3C286 after polcal dose not match to setjy parameter

### ISSUE THREAD

**Issue by Rsikezudta** [caracal-pipeline/caracal#1604]
Use '3C286' for xcal while excute worker polcal, the polarization angle(PA) is not correct.
'setjy' from the log file had set the polarization information of 3C286 to the model seems no problems, but the CORRECTED_DATA plot by worker inspect show the PA of 3C286 are 0 not 33 deg (0.575959 rad). 
I also check the MODEL_DATA plot using inspect, the PA plot are 0 deg not 33 deg.

Meanwhile, I imaged the calibrator 3C286 after polcal, but the PA was 28 deg, not 0, Did not match with 'setjy' 33 deg.
If there are some problems, how can I fix it?

Here to show a part of logs, plots and image.

	INFO	setjy::::+	##########################################
	INFO	setjy::::+	##### Begin Task: setjy              #####
	INFO	setjy::::	setjy(vis="/stimela_mount/msdir/1529521526_sdp_l0-cal.ms",field="3C286",spw="",selectdata=True,timerange="",
	INFO	setjy::::+	        scan="",intent="",observation="",scalebychan=True,standard="manual",
	INFO	setjy::::+	        model="",modimage="",listmodels=False,fluxdensity=[14.7172],spix=[-0.4507, -0.1798, 0.0357],
	INFO	setjy::::+	        reffreq="1.47GHz",polindex=[0.098],polangle=[0.575959],rotmeas=0.0,fluxdict={},
	INFO	setjy::::+	        useephemdir=False,interpolation="linear",usescratch=True,ismms=False)
	INFO	setjy::::	{'field': '3C286'}
	INFO	Imager::open() 	Opening MeasurementSet /stimela_mount/msdir/1529521526_sdp_l0-cal.ms
	INFO		Clearing all model records in MS header.
	INFO	setjy::::	CASA Version 5.8.0-109
	INFO	setjy::::
	INFO	imager::setjy()	Using channel dependent flux densities
	INFO		Clearing all model records in MS header.
	INFO	imager::data selection	Selected 135420 out of 2620560 rows.
	INFO		Clearing all model records in MS header.
	INFO	imager::data selection	Selected 135420 out of 2620560 rows.
	INFO	imager::ft() 	Fourier transforming: replacing MODEL_DATA column
	INFO	imager::createSkyEquation() 	Processing after subtracting componentlist /stimela_mount/msdir/1529521526_sdp_l0-cal.ms_setjy_spw0_3C286_0.856GHz58289.8d.cl
	INFO	imager::createFTMachine() 	Performing interferometric gridding...
	INFO	imager::setjy()	Flux density as a function of frequency (channel 0 of each spw):
	INFO	imager::setjy()+	  Frequency (GHz)    Flux Density (Jy, Stokes I)
	INFO	imager::setjy()	     0.856         18.3354
	INFO	setjy::::	##### End Task: setjy                #####
	INFO	setjy::::+	##########################################

CORRECTED_DATA phase-freq
![1gc-1529521526_sdp_l0-cal-3C286-CORRECTED_DATA-XY-YX-phase-FREQ-ANTENNA1](https://github.com/user-attachments/assets/2d059692-5ad4-4bc2-b195-b21b596d63ce)

MODEL_DATA
![1gc-1529521526_sdp_l0-cal-3C286-MODEL_DATA-XX-XY-YX-YY-phase-FREQ-ANTENNA1](https://github.com/user-attachments/assets/c9fc5032-57b4-41f4-9743-8a6dec266317)

Image of 3C286 after polcal by CARTA
![image](https://github.com/user-attachments/assets/2170e50b-c633-4143-b7f7-501ffb71f458)

**Comment by Athanaseus** [comment]
@Rsikezudta thanks for reporting this.

There have been recent updates to `caracal`. I'm curious, which version are you currently running?
Please also provide the full log as this may help us narrow down the problem.

Kind regards

**Comment by Rsikezudta** [comment]
Thanks for your reply. 

Caracal was executed on ilifu serverion and which container I used is STIMELA_IMAGES_1.7.7. 
Also following the log, the caracal version is 1.1.1. 
Here is the full log file. 
Should I change to recent version?
[slurm-9906533.zip](https://github.com/user-attachments/files/17597689/slurm-9906533.zip)

Thanks in advance

**Comment by Athanaseus** [comment]
Hi @Rsikezudta,

It may be worthwhile to run with the latest version.
However, upon reviewing the corrected data and the log, it appears that some spectral channels were not flagged properly.
The phase scatter is likely caused by unflagged RFI in those channels.

You can address this by enabling `flag_spw` for the calibrator and
specifying the problematic frequencies, including band roll-offs (note that the end channels are not inclusive):

```
flag:
  flag_spw:
    enable: true
    chans: '*:856~880MHz , *:1658~1800MHz, *:1419.8~1421.3MHz'
    ensure_valid: false
```

Out of curiosity, could you also share the bandpass gain solutions that were applied (e.g., .B1)?

Hope this helps!

Best regards

**Comment by Rsikezudta** [comment]
The RFI channels may very useful for me, thank you.

Here I show the gain solutions B1 which were applied.
![mypipelinerun-1529521526_sdp_l0-1kgb_primary B1](https://github.com/user-attachments/assets/0a696ee7-76a5-47d5-a588-27da729b0aa6)

Best regards

**Comment by AnnalisaB** [comment]
Hi all, my colleagues and I are having the same issue on several datasets where 3C286 is observed. We almost never recover the model we specify. After some reasearh, we have found this memo, where there's a lot of useful info: 
https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://archive-gw-1.kat.ac.za/public/repository/10.48479/bqk7-aw53/data/Absolute_linear_polarization_angle_calibration_using_planetary_bodies_for_MeerKAT_and_JVLA-REV-C.pdf&ved=2ahUKEwjW3MKo9MyJAxVMlP0HHQWHAx4QFnoECBkQAQ&usg=AOvVaw1sjkLIB0z8gfCH2pfAIBYg

Using this model, a PhD student, Alessandro Benati, manages to obtain a good match in UHF. He has 3C286 observed during night-time.
I don't have the same luck with my L band observations, where 3C286 is observed during day-time.

**Comment by a-benati** [comment]
> Hi all, my colleagues and I are having the same issue on several datasets where 3C286 is observed. We almost never recover the model we specify. After some reasearh, we have found this memo, where there's a lot of useful info: 
> 
> https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://archive-gw-1.kat.ac.za/public/repository/10.48479/bqk7-aw53/data/Absolute_linear_polarization_angle_calibration_using_planetary_bodies_for_MeerKAT_and_JVLA-REV-C.pdf&ved=2ahUKEwjW3MKo9MyJAxVMlP0HHQWHAx4QFnoECBkQAQ&usg=AOvVaw1sjkLIB0z8gfCH2pfAIBYg
> 
> 
> 
> Using this model, a PhD student, Alessandro Benati, manages to obtain a good match in UHF. He has 3C286 observed during night-time.
> 
> I don't have the same luck with my L band observations, where 3C286 is observed during day-time. 
> 
> 

To expand on this, in UHF band the model of 3C286 of that memo predicts:
- pol angle: 23.8 deg
- pol fraction: 6.4%

And I obtain:
- pol angle: 24.2 deg
- pol fraction: 6.6%

While using other models, such as the one implemented in CARACal, I'm not able to recover the values of the model that are specified.

**Comment by bennahugo** [comment]
Have you done ionospheric corrections? There can be up to -6 rad / m2
offsets

**Comment by a-benati** [comment]
@bennahugo no, I didn't apply any ionospheric corrections to my data, while in the model of the memo, if I understood correctly, they are applied before performing the fit, right?

**Comment by bennahugo** [comment]
You need to at least do a per scan correction of both your calibrators and
targets to account for diurnal effects. You can get the highest cadence
measurements using ALBUS

**Comment by AstroRipples** [comment]
Hi everyone,

Just to throw my 0.02€ into the discussion here on 3C286, I thought I'd share my results for 3C286 // J1331+3030 from an observing block I've been working on. 

Below you can see the values measured on the images given by running WSclean in full polarisation on the calibrated data:

![MeerKAT_Lband_Polarisation_J1331+3030_CBID-1639795398](https://github.com/user-attachments/assets/18b9b3e6-35b6-4b3a-ac44-70db7c348ac3)

The left panel shows the Stokes I flux density (left-hand axis) and linearly-polarised flux density (right-hand axis), centre panel shows the polarisation fraction, and right panel shows the polarisation angle, measured in 48 channels across the MeerKAT L-band. The orange datapoints show the values reported in Taylor & Legodi (2024), whereas the shaded region and horizontal line show the values reported in Table 1 of the [MeerKAT polarisation calibration wiki page](https://skaafrica.atlassian.net/wiki/spaces/ESDKB/pages/1493631000/Polarisation+calibration).

For this observation (CBID 1639795398) in 2021, we had two 5-minute scans on 3C286, separated by some parallactic angle. I've processed the data using `CARACal` following the standard calibration procedure. Assuming I recall the `CARACal` workflow correctly, as we have two scans, the `polcal` worker does not attempt to set a model, but derives the polarisation properties using the observed gain tables.

The fractional polarisation we get is nicely consistent with what's expected for 3C286, although there's clearly an offset in the polarisation angle. I haven't looked into this much further yet, as I've been primarily concerned with other datasets where I have 3C138 (I'll open a new issue to discuss that source later this afternoon) but as this observing run took place during early daylight hours (sunrise at 03:30 UTC, pol cal scans circa 06:00 and 08:36 UTC) I suspect that ionospheric RM is the culprit. A quick back-of-the-envelope calculation suggests an ionospheric RM of ~2 to ~3 rad/m2 would give rise to the observed offset with respect to the values reported by Taylor & Legodi. 

Will report back with more once I've had the chance to run `RMextract` ...

**Comment by ErikOsinga** [comment]
Hi all, I also have a couple cents to add to the discussion. I've been processing L-band data with MeerKAT as well, and my results are also a bit off from the most recent model. However, by digging through the caracal logs I found that it's actually calibrating against what I call the "historic" model in polcal, (i.e. EVPA angle 33 deg, no RM, polfrac = 0.098, no depolarisation)

![Image](https://github.com/user-attachments/assets/8ffbdb2d-2975-4b4e-8fb8-dab7d2576876)


![Image](https://github.com/user-attachments/assets/994f0fde-0239-487f-9a10-e3d9d8ecc840)


However, the corrected data is in closer agreement to the updated (actual) model of 3C286 than the historic model, but there's still an offset of a few degrees in polarisation angle. I wonder if this could be an ionospheric effect, because the residuals do seem to be approximately linear with wavelength squared:
(Note that this "correction" is simply an **inferred ionosphere**, I haven't looked into actual TEC content of that day. )



![Image](https://github.com/user-attachments/assets/7e61ab5f-780d-4e44-b051-7af7c6c5e8e5)





But my question is: wouldn't the ionospheric RM be taken out of the data (when using the correct model) when calibrating `Xf` in casa? (which is one of the steps in the caracal pipeline).

Additionally, are there plans to update the model to the improved model from B. Hugo & R. Perley + 2024?

**Comment by AnnalisaB** [comment]
Dear all,

I suggest to look at the results on the scans of the calibrator separately
Having an initial model for the caibrator in is not even necessary, though it helps
There are ionospheric effects, but the correction can be done a posteriori once the data are calibrated
i.e. you calibrate and then once you’re checking you add the RM that RMextract is returning and the data 
fit to the model, that’s what I’ve seen of 7 targets now (one example in attach)

What I do now, is looking at the scans of the calibrator separately, and picking the best scan to calibrate
(less RFI/higher elevation)
I’ve seen it helps
Not sure caracal allows that (I’m doing the reduction in CASA) but I know Francesca is on it

Hope it helps

Cheers
Annalisa

<img width="432" alt="Image" src="https://github.com/user-attachments/assets/f8610483-0fb8-49e6-a300-ea9a8f7509f0" />

**Comment by bennahugo** [comment]
Yup - to add to this Xf calibrates the crosshand phase. In linear basis this just removes the elliptical polarization induced by the system - it does not set any angle (unlike in circular basis where the crosshand phase is just a further rotation between U and Q). The absolute angle is set by the feed alignment on sky (which is << 2 degrees on most receivers at all bands from our work looking at the Moon as reference calibrator).

As for the ionosphere (and adding to the reason above regarding angles) - no calibration against 3C286 does not correct the ionosphere. It is a spatio-temporal effect that can rapidly change between daytime and nighttime and within daytime depending solar zenith. It will be different depending on where you pierce the ionosphere and at what time.

You need to use AIPS TECOR / RMExtract / ALBUS with GNSS data from the TrigNET stations around MeerKAT site (Calvinia, De Aar and Sutherland) to correct for the variable TEC towards your target and calibrators separately.

**Comment by AstroRipples** [comment]
Hi fellow MeerKAT polarisation enthusiasts 👋  documenting my experience here:

I've been experimenting with @AnnalisaB 's recipe for calibrating 3C286 in CASA, and comparing with a run using the CARACal version I have installed in my virtual environment (which may well not be the latest). Similar to others on this thread, in the case of this dataset (L-band, CBID 1618694918) I measure a lower polarisation fraction and polarisation angle when imaging data calibrated in CARACal than when calibrating the same dataset (including the same flagging) in CASA:

![Image](https://github.com/user-attachments/assets/44d50bf6-4985-4e93-b67e-fcbca8529b58)

The polarisation angle before correction for ionospheric RM is shown here as the "(Observed)" datapoints (triangles) whereas the RM-corrected version is shown by the circles. Using `RMextract` for the ionospheric correction I recover a polarisation angle that's pretty close to the model from Hugo & Perley and the observed results from Taylor & Legodi for the CASA-calibrated data, a little further (although still pretty close) for the CARACal-calibrated data. 

In this case, I have two scans of 3C 286 which I'm imaging together and applying the average ionospheric RM correction to the observed measurements, which may explain why there's a slight offset. Overall though, these results are within 2 degrees of the expected behaviour, which is satisfying. The CASA-calibrated data also gives a polarisation fraction that is much closer to both the H&P model and T&L observations.
